src/snapshot.rs

```diff
@@ -0,0 +1,76 @@
+use crate::instructions::Register;
+use crate::memory::Memory;
+use crate::memory::FLAG_DIRTY;
+use crate::{
+    CoreMachine, Error, RISCV_GENERAL_REGISTER_NUMBER, RISCV_PAGES, RISCV_PAGESIZE,
+    RISCV_PAGE_SHIFTS,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Default, Deserialize, Serialize)]
+pub struct Snapshot {
+    pub version: u32,
+    pub registers: [u64; RISCV_GENERAL_REGISTER_NUMBER],
+    pub pc: u64,
+    pub page_indices: Vec<u64>,
+    pub pages: Vec<Vec<u8>>,
+}
+
+pub fn make_snapshot<T: CoreMachine>(machine: &mut T) -> Result<Snapshot, Error> {
+    let mut snap = Snapshot::default();
+    snap.version = machine.version();
+    for (i, v) in machine.registers().iter().enumerate() {
+        snap.registers[i] = v.to_u64();
+    }
+    snap.pc = machine.pc().to_u64();
+
+    for i in 0..RISCV_PAGES {
+        let flag = machine.memory_mut().fetch_flag(i as u64)?;
+        if flag & FLAG_DIRTY != 0 {
+            let addr_from = i << RISCV_PAGE_SHIFTS;
+            let addr_to = (i + 1) << RISCV_PAGE_SHIFTS;
+
+            let mut page = vec![0; RISCV_PAGESIZE];
+            for i in (addr_from..addr_to).step_by(8) {
```

Comment:
Can we add some comments to document how snapshot works?

Reply:
I submitted an update, and now the comments will explain what we did.

```rs
// Snapshot provides a mechanism for suspending and resuming a virtual machine.
//
// When cycle limit is too low, our work won't be finished when we hit it and
// a "max cycles exceeded" error will be returned. At this time, a snapshot could
// be created, we can create a new virtual machine from this snapshot, provide
// more cycles and then continue to run it.
//
// For the following data, we simply save them, and then restore them.
//   - machine.version
//   - machine.pc
//   - machine.registers
//
// For memory, the situation becomes more complicated. Every memory page has
// page flag where each page flag stores a optional FLAG_DIRTY. When this page
// is wrote, the memory instance will set this bit to 1, this helps us keep
// track of those pages that have been modified. After `machine.load_elf`, We
// clean up all dirty flags, so after the program terminates, all pages marked
// as dirty are the pages that have been modified by the program. We only store
// these pages in the snapshot.
```